src/ocean/util/container/btree/BTree.d

```diff
@@ -0,0 +1,695 @@
+            debug (BTreeSanity) check_invariants(*this);
+            if (ret)
+                this.content_version++;
+            return ret;
+        }
+        else
+        {
+            auto ret = this.impl.insertNonFull(this.root, el);
+            debug (BTreeSanity) check_invariants(*this);
+            if (ret)
+                this.content_version++;
+            return ret;
+        }
+    }
+
+    /******************************************************************************
+
+        Deletes the element from the BTree.
+
+        Params:
+            element = element to delete.
+
+        Returns:
+            true if the element was found and deleted, false otherwise.
+
+     ******************************************************************************/
+
+    public bool removeElement (ref ElementType element)
+    {
+        BTreeNode* parent = null;
+
+        bool rebalance_parent;
+        auto res = deleteFromNode(this.root,
+                parent, element, rebalance_parent);
+
+        // can't rebalance the root node here, as they should all be
+        // rebalanced internally by deleteFromNode
+        assert(rebalance_parent == false);
+
+        debug (BTreeSanity) check_invariants(*this);
+
+        if (res)
+            this.content_version++;
+
+        return res;
+    }
+
+    /******************************************************************************
+
+        Finds the highest value in the tree.
+
+        Returns:
+            pointer to largest element in the tree, or null if the tree is empty.
+
+    ******************************************************************************/
+
+    public Const!(ElementType)* findMaximum ()
+    {
+        size_t index;
+
+        if (this.root is null)
+        {
+            return null;
+        }
+
+        return &(this.impl.findMaximum(this.root, index).elements[index]);
+    }
+
+    /******************************************************************************
+
+        Searches the tree for the element and returns the pointer to it, or null
+        if not found.
+
+        Params:
+            tree = tree to search in
+            element = element to find in a tree.
+
+        Returns:
+            pointer to the element, or null if not found.
+
+    *******************************************************************************/
+
+    public Const!(ElementType)* search (ref ElementType element)
+    {
+        size_t index;
+        auto node = this.impl.search(element, index);
+
+        return node is null? null : &node.elements[index];
+    }
+
+    /**************************************************************************
+
+        Node of the tree. Contains degree * 2 - 1 elements and degree * 2
+        subtrees.
+
+    **************************************************************************/
+
+    package struct BTreeNode
+    {
+        /// Number of the elements currently in the node
```

Comment:
This contradicts the statement above "Contains degree * 2 - 1 elements". Should that say "at most"?